src/main/scala/com/gu/invoicing/refund/Impl.scala

```diff
@@ -0,0 +1,246 @@
+      .pipe(read[InvoiceItemQueryResult](_))
+      .records
+      .groupBy(_.InvoiceId)
+  }
+
+  def getInvoicePaymentId(invoiceId: String): Option[String] = {
+    HttpWithLongTimeout(s"$zuoraApiHost/v1/action/query")
+      .header("Authorization", s"Bearer ${accessToken()}")
+      .header("Content-Type", "application/json")
+      .postData(s"""{"queryString": "select Id, invoiceId, paymentId from InvoicePayment where invoiceId = '$invoiceId'"}""")
+      .method("POST")
+      .asString
+      .body
+      .pipe(read[InvoicePaymentQueryResult](_))
+      .records
+      .headOption
+      .map(_.PaymentId)
+    //      .getOrElse(throw new Exception(s"There should be one invoice payment against invoice $invoiceId"))
+  }
+
+  def createRefundObject(amount: Double, paymentId: String, comment: String): String = {
+    HttpWithLongTimeout(s"$zuoraApiHost/v1/object/refund")
+      .header("Authorization", s"Bearer ${accessToken()}")
+      .header("Content-Type", "application/json")
+      .postData(
+        s"""
+           |{
+           |  "Amount": $amount,
+           |  "Comment": "$comment",
+           |  "PaymentId": "$paymentId",
+           |  "Type": "Electronic"
+           |}
+           |""".stripMargin)
+      .method("POST")
+      .asString
+      .body
+      .pipe(body => read[RefundResult](body))
+      .Id
+  }
+
+  def getRefundStatus(refundId: String): String = {
+    HttpWithLongTimeout(s"$zuoraApiHost/v1/object/refund/$refundId")
+      .header("Authorization", s"Bearer ${accessToken()}")
+      .asString
+      .body
+      .pipe(read[Refund](_))
+      .Status
+  }
+
+  def netAdjustmentsByInvoiceItemId(adjustments: List[InvoiceItemAdjustment]): Map[String, Double] = {
+    adjustments
+      .groupBy(_.SourceId)
+      .map { case (invoiceItemId, adjustments) =>
+        val netCredits = adjustments.filter(_.Type == "Credit").map(_.Amount).sum
+        val netCharges = adjustments.filter(_.Type == "Charge").map(_.Amount).sum
+        (invoiceItemId, netCredits - netCharges)
+      }
+  }
+
+  /**
+   * This is likely the most complicated part of the program. It decides which invoice items to adjust and
+   * by how much by taking into account any previous adjustments already made to corresponding items.
+   */
+  def spreadRefundAcrossItems(
+    invoiceItems: List[InvoiceItem],
+    adjustments: List[InvoiceItemAdjustment],
+    totalRefundAmount: Double,
+    refundGuid: String,
+  ): List[InvoiceItemAdjustmentWrite] = {
+
+    /* Collect all item adjustments of a particular invoice item and return remaining amount that can be adjusted/refunded */
+    def availableAmount(invoiceItem: InvoiceItem): Option[Double] = {
+      netAdjustmentsByInvoiceItemId(adjustments).get(invoiceItem.Id) match {
+        case Some(netAdjustment) =>
+          val availableRefundableAmount = invoiceItem.ChargeAmount - netAdjustment
+          if (availableRefundableAmount <= 0.0) None else Some(availableRefundableAmount)
```

Comment:
```suggestion
          if (availableRefundableAmount <= 0) None else Some(availableRefundableAmount)
```

Reply:
Addressed in https://github.com/guardian/invoicing-api/pull/2